src/rgw/librgw.cc

```diff
@@ -61,6 +62,11 @@
 
 bool global_stop = false;
 
+static void handle_sigterm(int signum)
+{
+  dout(20) << __func__ << " SIGUSR1 ignored" << dendl;
```

Comment:
proabably shouldn't ignore signum

Reply:
Hi Matt, here is another way to fix this problem: prevent to send SIGUSR1 by updating file src/rgw/rgw_lib_frontend.h: 
```c++
 void stop() override {
  RGWProcessFrontend::stop();  //remove this 
  get_process()->stop();
}
```
When get_process()->stop() is invoking, the gc loop in librgw.cc should break because shutdown == true:
```c++
void RGWLibProcess::run()
  {
    /* write completion interval */
    RGWLibFS::write_completion_interval_s =
      cct->_conf->rgw_nfs_write_completion_interval_s;

    /* start write timer */
    RGWLibFS::write_timer.resume();

    /* gc loop */
    while (! shutdown) {
      lsubdout(cct, rgw, 5) << "RGWLibProcess GC" << dendl;

      /* dirent invalidate timeout--basically, the upper-bound on
       * inconsistency with the S3 namespace */
      auto expire_s = cct->_conf->rgw_nfs_namespace_expire_secs;

      /* delay between gc cycles */
      auto delay_s = std::max(int64_t(1), std::min(int64_t(MIN_EXPIRE_S), expire_s/2));

      unique_lock uniq(mtx);
    restart:
      int cur_gen = gen;
      for (auto iter = mounted_fs.begin(); iter != mounted_fs.end();
	   ++iter) {
	RGWLibFS* fs = iter->first->ref();
	uniq.unlock();
	fs->gc();
	fs->rele();
	uniq.lock();
	if (cur_gen != gen)
	  goto restart; /* invalidated */
      }
      uniq.unlock();
      std::this_thread::sleep_for(std::chrono::seconds(delay_s));
    }
  }
```
However, this loop will be blocked at : 
```c++
 std::this_thread::sleep_for(std::chrono::seconds(delay_s));
```
**delay_s** is determined by **rgw_nfs_namespace_expire_secs**  in ceph.conf, which is set to 300s by default. If we'd like to shutdown librgw quickly, we have to set **rgw_nfs_namespace_expire_secs** as short as possible, which could bring a lower performance. So we have to improve the mechanism of blocking gc loop if we want to fix librgw shutdown err in this way.

Reply:
we don't want to do that;  we could change the sleep to be a timedwait on a semaphore;  that's how I've dealt with this sort of thing in the past

Reply:
I agree with that :)